__tests__/compiler.spec.ts

```diff
@@ -0,0 +1,853 @@
+            for (let error in output.errors) {
+                error = output.errors[error];
+                if (error['type'] === errorType) {
+                    if (message) {
+                        if (error['message'].match(message) !== null) {
+                            return true;
+                        }
+                    } else {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    function expectNoError(output: any) {
+        if (output['errors']) {
+            for (let error in output['errors']) {
+                error = output['errors'][error];
+                if (error['severity'] === 'error') {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+
+    describe(versionText, () => {
+
+        describe('Version and license', () => {
+            test('check version', () => {
+                expect(typeof solc.version()).toBe('string');
+            });
+            test('check semver', () => {
+                expect(typeof solc.semver()).toBe('string');
+            });
+            test('check license', () => {
+                expect(typeof solc.license() === 'undefined' || typeof solc.license() === 'string').toBeTruthy();
+            });
+
+        });
+    })
+
+    describe('Compilation', () => {
+        test('single files can be compiled (using lowlevel API)', () => {
+            if (typeof solc.lowlevel.compileSingle !== 'function') {
+                console.log('Low-level compileSingle interface not implemented by this compiler version.');
+                return
+            }
+
+            let output = JSON.parse(solc.lowlevel.compileSingle('contract A { function g() public {} }'));
+            expect(output['contracts']).toBeTruthy();
+            let bytecode = getBytecode(output, '', 'A');
+            expect(typeof bytecode).toBe('string');
+            expect(bytecode.length).toBeGreaterThan(0);
+        });
+
+        /* TODO: fixme      
+        test('invalid source code fails properly (using lowlevel API)', () => {
+          // TODO: try finding an example which doesn't crash it?
+          if (semver.eq(solc.semver(), '0.4.11')) {
+            console.log('Skipping on broken compiler version');
+            return;
+          }
+  
+          if (typeof solc.lowlevel.compileSingle !== 'function') {
+            console.log('Low-level compileSingle interface not implemented by this compiler version.');
+            return;
+          }
+  
+          let output = JSON.parse(solc.lowlevel.compileSingle('contract x { this is an invalid contract }'));
+          if (semver.lt(solc.semver(), '0.1.4')) {
+            expect(output.error.indexOf('Parser error: Expected identifier')).not.toBe(-1);
+            return;
+          }
+
+          // TODO: what plan does? https://github.com/substack/tape#tplann
+          //st.plan(3);
```

Comment:
@chriseth does this mean that tests run 3 times?
https://github.com/substack/tape#tplann 

Definition on the github:
"Declare that n assertions should be run. t.end() will be called automatically after the nth assertion. If there are any more assertions after the nth, or after t.end() is called, they will generate errors."

Reply:
Yes

Reply:
Actually no, it says that there are 3 checks to be run. Because they are async, you basically tell the test framework when to stop waiting for more checks.